src/Workspaces/Core/MSBuild.BuildHost/Rpc/RpcServer.cs

```diff
@@ -0,0 +1,146 @@
+        private volatile int _nextRpcTargetIndex = -1; // We'll start at -1 so the first value becomes zero
+
+        public RpcServer(Stream sendingStream, Stream receivingStream)
+        {
+            _sendingStream = new StreamWriter(sendingStream, Encoding.UTF8);
+            _receivingStream = new StreamReader(receivingStream, Encoding.UTF8);
+        }
+
+        public int AddTarget(object rpcTarget)
+        {
+            // Loop until we successfully have a new index for this; practically we don't expect this to ever collide, since that'd mean we'd have
+            // billions of long lived projects, but...
+            while (true)
+            {
+                var nextIndex = Interlocked.Increment(ref _nextRpcTargetIndex);
+                if (_rpcTargets.TryAdd(nextIndex, rpcTarget))
+                    return nextIndex;
+            }
+        }
+
+        public async Task RunAsync()
+        {
+            var runningTasks = new ConcurrentSet<Task>();
+
+            string? line;
+            while ((line = await _receivingStream.ReadLineAsync().ConfigureAwait(false)) != null)
+            {
+                var request = JsonConvert.DeserializeObject<Request>(line);
+
+                Contract.ThrowIfNull(request);
+
+                var runningTask = Task.Run(async () =>
+                    {
+                        Response response;
+
+                        try
+                        {
+                            Contract.ThrowIfFalse(
+                            _rpcTargets.TryGetValue(request.TargetObject, out var rpcTarget),
+                            $"Received a request for target object {request.TargetObject} but we don't have a registered object for that.");
+
+                            var method = rpcTarget.GetType().GetMethod(request.Method, BindingFlags.Public | BindingFlags.Instance);
+
+                            Contract.ThrowIfNull(method, $"The invoked method '{method}' could not be found.");
+
+                            var methodParameters = method.GetParameters();
+                            var arguments = new object?[methodParameters.Length];
+
+                            for (var i = 0; i < methodParameters.Length; i++)
+                            {
+                                // If the method we're calling accepts a cancellation token, we wouldn't have passed that, so add in a CancellationToken.None here. Although we could
+                                // remove the cancellation from the underlying method, this keeps that support around should we need it.
+                                if (i == methodParameters.Length - 1 && methodParameters[i].ParameterType == typeof(CancellationToken))
+                                    arguments[i] = CancellationToken.None;
+                                else
+                                    arguments[i] = request.Parameters[i].ToObject(methodParameters[i].ParameterType);
```

Comment:
we're blindly indexing into the request paramteers.  can you do some more validation of htat up front so we can give good error messages?

Reply:
Extra validation added.